### PR TITLE
Fix null.isInstanceOf

### LINKED ISSUE #283: null.{isInstanceOf, asInstanceOf} semantics differs
Currently all operations on `null`s are undefined behaviour. This is not consistent with `isInstanceOf` and `asInstanceOf` having a defined non-throwing semantics on null values on the JVM.

### PR DESCRIPTION
**Pull request by pbatko**
In `preBlock` looks for `Op.Is` in a `Block` and adds runtime check if the operand is `null`.
In `preInst`, if operand is the `null` literal returns `false`, otherwise goes to proper `instanceOf` checking.

Limitations: 
- doesn't work yet - tests fail at runtime,
- is there an easier way add an `if\else` expression then splitting and creating `Block`s manually?
- static check for null literal in `preInst` is useless as `preBlock` introduces runtime check earlier.

Fixes #283

### CODE REVIEW

**Comment by densh** [comment]
> is there an easier way add an if\else expression then splitting and creating Blocks manually?

This is one is major flaw in the way we structure NIR. Currently function's code is a sequence of blocks that contain sequences of static assignments and terminating control-flow instruction. Due to this it's indeed very inconvenient to write a pass that introduces new control-flow.

Luckily there is a simple solution: just make basic blocks implicit. I've been working on this for past two days in [`topic/implicit-basic-blocks`](https://github.com/scala-native/scala-native/tree/topic/implicit-basic-blocks). Function's code is a one-level sequence of instructions now. Blocks are demarcated by `Label` instruction. Because of the flat structure, introducing new control-flow becomes trivial. It complicates some other a bit, but the trade-off is worth it. There is light in the end of the tunnel.

**Comment by sjrd** [comment]
> if operand is the null literal returns true

Attention, this is not correct. When `x eq null`, the results should be:
* `x.isInstanceOf[A]` is `false` for all `A`
* `x.asInstanceOf[A]` *succeeds* for all `A` (even though `isInstanceOf[A]` returns false)

**Comment by pbatko** [comment]
@sjrd  Thanks for noticing!

@densh Great. When are you planning to merge implicits blocks into master? 
Also, I think I got `isInstanceOf` working now and will probably take a look at `asInstanceOf` now.

Btw. AFAIK when using LLVM as the backend, a compiler could completely do away with basic blocks and SSA and use `alloca` followed by LLVM's `mem2reg` pass. Why was it chosen not to do that?

**Comment by densh** [comment]
@pbatko 

Implicit blocks nearly done, maybe a few more days of work to be on the safe side.

Speaking of basic blocks: both scalac and dotty have a concept of labels that are generated for pattern matching, loops and tail recursive code. The easiest way to support that is to have basic blocks ourselves. Moreover, basic-blocks are quite nice if you want to do optimisations on the NIR. (That's what @hkapp and @Duhemm are looking at right now.)

**Comment by densh** [comment]
@pbatko 

This PR will need to be updated as soon as #336 is merged. The fix should be way easier with implicit blocks.

**Comment by pbatko** [comment]
@densh Ok.

**Comment by densh** [comment]
Implicit basic blocks have just been merged! It's time to rebase this one.

File: tools/src/main/scala/scala/scalanative/compiler/pass/IsLowering.scala [hunk 1]
```diff
@@ -8,7 +8,44 @@ import nir._, Inst.Let
...
+          Let(isNullV.name,
+              Op.Comp(Comp.Ieq, Type.Ptr, obj, Val.Zero(Type.Ptr))),
+          Inst.If(isNullV, Next(thenL), Next(elseL)),
+          // then
+          Inst.Label(thenL, Seq.empty),
+          Let(thenResultV.name, Op.Copy(Val.False)),
+          Inst.Jump(Next.Label(contL, Seq(thenResultV))),
+          // else
+          Inst.Label(elseL, Seq.empty)) ++
+      doEliminateIs(Let(elseResultV.name, Op.Is(ty, obj))) ++
```
Comment (pbatko) [inline tools/src/main/scala/scala/scalanative/compiler/pass/IsLowering.scala]: this list concatention doesn't look very pretty..
maybe it'd better if this pass looked like this:
`override def preInst = case a => handleNull(a).flatMap(eliminateIs)`?

File: unit-tests/src/main/scala/scala/scalanative/native/InstanceOfSuite.scala [hunk 2]
```diff
@@ -0,0 +1,79 @@
+package scala.scalanative.native
+
+object InstanceOfSuite extends tests.Suite {
+
+  // asInstanceOf
+
+//  test("expects anyRef.asInstanceOf[String] fail") {
```
Comment (pbatko) [inline unit-tests/src/main/scala/scala/scalanative/native/InstanceOfSuite.scala]: now `(new AnyRef).asInstanceOf[String]` succeeds
what should be the result here?
is it success in case of prod mode and failure in case of dev mode?

Comment (densh) [inline unit-tests/src/main/scala/scala/scalanative/native/InstanceOfSuite.scala]: That's good, `asInstanceOf` has a bunch of other corner cases we don't fully handle right now, apart from null. I'm going to look at this today. This definitely should go into a separate PR.

File: tools/src/main/scala/scala/scalanative/compiler/pass/IsLowering.scala (hunk 1 above)
Comment (densh) [inline tools/src/main/scala/scala/scalanative/compiler/pass/IsLowering.scala]: I don't think that it will really change much. The current solution looks good to me.

**Comment by densh** [comment]
LGTM, thanks a lot @pbatko ! 🍰
